[browser] wasm memory snapshot into browser cache

pavelsavara opened the pull request:
- create memory snapshot after first load
- skip downloads of assets when there is memory already in cache
- resize wasm memory as necessary and apply the cached memory bytes
- calculate hash for all assets and the config file
- added new `.withStartupMemoryCache()` public API. `true` by default.
- start threads and eventPipe bit later (after taking snapshot)
- disabled JITERP during creation of snapshot.

## Other
- cleanup of `anyModule`
- moved ICU shard logic to `icu.ts`

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -457,20 +465,37 @@ async function instantiate_wasm_module(
+            const wasmMemory = anyModule.asm.memory;
+            // .grow() takes a delta compared to the previous size
+            wasmMemory.grow((memoryBytes!.byteLength - wasmMemory.buffer.byteLength + 65535) >>> 16);
+            runtimeHelpers.updateGlobalBufferAndViews(wasmMemory.buffer);
```
@radekdoulik I will need to tell emscripten that I resized the memory. 
So the `replacements.updateGlobalBufferAndViews` should survive you PR in some form.

pavelsavara committed:
wip

pavelsavara committed:
wip

pavelsavara committed:
wip

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

# Conflicts:
#	src/mono/wasm/runtime/types.ts

kg commented:
Is this being turned on by default?

kg commented on src/mono/wasm/runtime/icu.ts:
```diff
@@ -50,3 +50,29 @@ export function mono_wasm_globalization_init(): void {
+        return icuAssets[0].name;
+
+    // reads the browsers locale / the OS's locale
+    const preferredCulture = ENVIRONMENT_IS_WEB ? navigator.language : Intl.DateTimeFormat().resolvedOptions().locale;
```
what happens if preferredCulture changes after the snapshot is taken?

pavelsavara replied:
the hash will change and we will load everything again

kg commented on src/mono/wasm/runtime/storage.ts:
```diff
@@ -0,0 +1,97 @@
+
+export async function storeMemory(configurationHash: number, memory: ArrayBuffer) {
+    try {
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${configurationHash}`;
```
Is GitHash actually sufficient? If the source they built the application from has local changes, the git hash will still be that of the most recent commit, perhaps with an -m on the end. Which means if they make more changes and run again, it could use an outdated memory cache, right?

pavelsavara replied:
There are hashes of .wasm file as well as all DLLs and assets in the `configurationHash` via mono-config.json `assetsHash`

pavelsavara replied:
`GitHash` is probably not even necessary

kg replied:
If we are including the assets hash that's great, the name "configuration hash" made me think it was configuration (like locale and timezone) only

pavelsavara replied:
I'll rename it and comment more

kg commented on src/mono/wasm/runtime/strings.ts:
```diff
@@ -276,3 +276,13 @@ export function js_string_to_mono_string_new(string: string): MonoString {
     }
 }
+
+export function get_hash_code(string: string) {
```
this probably needs a comment explaining the existence of a special purpose hashing routine when we could use the hashes provided by the browser or the managed runtime instead

kg replied:
personally, i would prefer to use a strong hash since a hash collision could result in loading the wrong memory

pavelsavara replied:
ok, I will use `crypto` API

pavelsavara committed:
wip

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

kg commented on src/mono/wasm/runtime/storage.ts:
```diff
@@ -0,0 +1,126 @@
+    } catch {
+        // There's no known scenario where we should get an exception here, but considering the
+        // Chromium bug above, let's tolerate it and treat as "proceed without caching".
+        return null;
```
We should log a message here

kg commented on src/mono/wasm/runtime/strings.ts:
```diff
@@ -275,4 +275,4 @@ export function js_string_to_mono_string_new(string: string): MonoString {
         temp.release();
     }
-}
+}
```
whitespace damage

kg reviewed (approved):
Everything in this PR looks good to me, BUT, I am uncomfortable enabling it by default unless we have a good .md file easily located in the repo that explains:
* How it works / what it does
* How to troubleshoot it
* How to disable it

For one example it's not very clear to me where precisely in the startup process the snapshot is taken. I know from reading the whole PR that it's early enough for certain assumptions to be made. Someone less familiar with the codebase will be at a total loss when it comes to that

I think we may also want to make sure we always log a message to the console when loading a snapshot that says we loaded a snapshot and where we loaded it from, not just when diagnostictracing is set. It's very important that if something really strange/bad happens due to memory snapshots, there is a little signal in the console that lets people easily rule out whether snapshots were involved.

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara commented:
>

maraf reviewed (approved):
I like it be on by default + I very much agree with @kg comment.

Should we update the benchmark, so the measure measurements are not affected by this change? What is the strategy there?

pavelsavara committed:
feedback

pavelsavara commented:
> it's not very clear to me where in the startup process the snapshot is taken. 
>     * How it works / what it does
>     * How to disable it

Good feedback. Added document.

> How to troubleshoot it

I don't know yet. Everything just works so far :-D

> I am uncomfortable enabling it by default 

Let's collect some experience first. We can disable it later in the release, if the feature is not good enough. 

> I think we may also want to make sure we always log a message to the console when loading a snapshot 

We don't do that in Release configuration, we are just silent about any configuration.

(Lot of these comments would apply to jiterpretter too)

kg commented:
> I like it be on by default + I very much agree with @kg comment.
> 
> Should we update the benchmark, so the measure measurements are not affected by this change? What is the strategy there?

It would be great if we could somehow have a snapshot version of the two appstart measurements alongside a non-snapshot one, but I'm not sure how easy that would be to do. Maybe a special URL parameter like `#monowasm-no-snapshot` and we pass it for the no-snapshot measurement and not for the snapshot one, so that we can compare the timings?

pavelsavara commented:
> > Should we update the benchmark, so the measure measurements are not affected by this change? What is the strategy there?
> 
> It would be great if we could somehow have a snapshot version of the two appstart measurements alongside a non-snapshot one

I will create separate PR for cold start benchmark and we could merge it before this one.

kg commented:
> > I think we may also want to make sure we always log a message to the console when loading a snapshot
> 
> We don't do that in Release configuration, we are just silent about any configuration.
> 

I understand the reasoning but IMO the value of having a .md file in the repo is that a `snapshot loaded from ..., see https://etc/snapshot.md` message in the console would send users to a good place if they experience a failure loading the snapshot or if the application fails right after it loads

Otherwise, they have to somehow know that snapshots are involved and then go find the .md file.

Maybe in all snapshot related failure scenarios enough will be in the callstack to lead them to it?

pavelsavara committed:
lint

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+    - the mono-config has field `assetsHash` which is summary SHA256 of all the assets.
+    - the configuration could be changed programmatically and se we calculate the hash at the runtime, just before taking snapshot. 
+- the snapshot is taken just after we initialize the Mono runtime
+    - cwraps initialized (they would be initialized again on subsequent start)
```
```suggestion
    - before cwraps are initialized (they would be initialized again on subsequent start)
```

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+    - after loading all the DLLs into memory.
+    - after loading ICU and timezone data
+    - after applying environment variables and other runtime options.
+    - before any worker threads init
```
```suggestion
    - before any worker threads initialization
```

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+    - after applying environment variables and other runtime options.
+    - before any worker threads init
+    - before any JavaScript interop initialization
+    - before any Managed code initialization
```
```suggestion
    - before any Managed code executes
```

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+    - the DLLs and other downloaded assets each have SHA256 which is used to validate 'integrity' by the browser.
+    - the mono-config has field `assetsHash` which is summary SHA256 of all the assets.
+    - the configuration could be changed programmatically and se we calculate the hash at the runtime, just before taking snapshot. 
+- the snapshot is taken just after we initialize the Mono runtime
```
Please, unify ends of lines (whether they have "." or not is up to you 🙂)

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+    - therefore we do not expect to store any application state in the snapshot.
+
+### How to opt out
+ - you can use new API `withMemoryCache(false)` to opt out from the feature.
```
```suggestion
You can turn this feature of by calling `withMemoryCache(false)` on [dotnet API](https://github.com/dotnet/runtime/blob/main/src/mono/wasm/runtime/dotnet.d.ts).
```

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+
+We take snapshot of WASM memory after first cold start at run time​.
+We store it on the client side in the browser cache.
+For subsequent start of the application with the same configuration and same assets, we could use the snapshot.
+Instead of downloading all the assets and doing the runtime startup again.
+Such subsequent start is significantly faster.
```
```suggestion
For subsequent runs with the same configuration and same assets, we use the snapshot 
instead of downloading everything again and doing the runtime startup again.
These subsequent starts are significantly faster.
```

maraf reviewed (commented):
Please use take all my suggestions as optional

pavelsavara committed:
more linear memory usage and recovery from failure of snapshot use

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
feedback

pavelsavara commented:
@lambdageek I start thread pool and eventPipe small bit later here. Please let me know if that seems an issue.

@kg @maraf thanks for feedback so far!

- I slightly renamed stuff for better clarity
- I implemented bit more linear usage of JS engine's memory. 
- I also added recovery from hypothetical failure during `wasmMemory.grow` which Kate hinted about for mobile devices. I tested the recovery locally. It involves delaying download of DLLs until we know if the loading of snapshot was OK.

I also plan to make changes to our unit test suite to (allow them to) run from snapshot.

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -386,10 +394,6 @@ async function mono_wasm_after_user_runtime_initialized(): Promise<void> {
                 }
             }
         }
-        // for Blazor, init diagnostics after their "onRuntimeInitalized" sets env variables, but before their postRun callback (which calls mono_wasm_load_runtime)
-        if (MonoWasmThreads) {
-            await mono_wasm_init_diagnostics();
-        }
```
Is this still true in the new code? Is `mono_wasm_before_user_runtime_initialized` after blazor's `onRuntimeInitialized` and before their `postRun`?

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -318,32 +323,28 @@ function mono_wasm_pre_init_essential(isWorker: boolean): void {
     Module.removeRunDependency("mono_wasm_pre_init_essential_async");
 }
 
-// runs just in non-blazor
```
Did this PR change when this function runs, or was the comment out of date?

pavelsavara replied:
Comment was out of date. I will take the cleanup to separate PR.

pavelsavara replied:
I move the cleanup stuff to separate PR https://github.com/dotnet/runtime/pull/83082

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -318,32 +323,28 @@ function mono_wasm_pre_init_essential(isWorker: boolean): void {
     Module.removeRunDependency("mono_wasm_pre_init_full");
 }
 
-// runs just in non-blazor
```
Did this PR change when this function runs, or was the comment out of date?

pavelsavara replied:
Out of date, moved to https://github.com/dotnet/runtime/pull/83082

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -318,32 +323,28 @@ function mono_wasm_pre_init_essential(isWorker: boolean): void {
         Module.removeRunDependency("mono_wasm_pre_init_essential");
 }
 
-// runs in both blazor and non-blazor
```
Was this comment incrorrect?

pavelsavara replied:
There is no special more for blazor anymore, but it's unrelated to this PR

lambdageek reviewed (commented):
I have no idea if moving the threads and eventpipe init code is ok or not.  

The threads seem like they're ok.  I assume tests would fail if it wasn't.

For eventpipe + Blazor it's probably not ok to move anything since the whole startup process is fragile.  The issue is that blazor sets environmnent variable values based on their config.  And eventpipe needs to read the updated environment variable values - but also it must be initialized before we run any of the runtime's C startup code.

pavelsavara commented:
> For eventpipe + Blazor it's probably not ok to move anything since the whole startup process is fragile. 

They don't have custom startup anymore. I will test the event-pipe on our sample to see if I broke it or not.

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

# Conflicts:
#	src/mono/wasm/runtime/assets.ts
#	src/mono/wasm/runtime/imports.ts
#	src/mono/wasm/runtime/managed-exports.ts
#	src/mono/wasm/runtime/marshal.ts
#	src/mono/wasm/runtime/startup.ts
#	src/mono/wasm/runtime/types.ts

pavelsavara committed:
start runtime twice for unit tests, so that the actual tests is running on snapshots

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara committed:
wip

pavelsavara committed:
continue loading non-snapshot assets

pavelsavara committed:
disable adding methods to __indirect_function_table before we take memory snapshot

pavelsavara committed:
rename startupMemoryCache

pavelsavara committed:
try to fix threads

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara committed:
whitespace

pavelsavara committed:
fix

pavelsavara committed:
fix storing memory

pavelsavara committed:
assert

pavelsavara committed:
- make snapshot hash more resilient

- silent dry run in unit test

pavelsavara committed:
fix

pavelsavara committed:
- renamed memorySnapshotSkippedOrDone

- asset loading feedback
- removed mono_wasm_load_icu_data from wasm imports

pavelsavara committed:
- rename loadedMemorySnapshot

 - drop mono_wasm_load_runtime_done
 - reorder startup

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara committed:
fix

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

# Conflicts:
#	src/mono/wasm/runtime/startup.ts

pavelsavara committed:
- internal config exitAfterSnapshot

- do not emit default pthreadPoolSize to mono-config
- prefer to propagate exit code from exception in set_exit_code_and_quit_now
- allow silent errors in err.silent in _print_error
- postpone mono_wasm_init_diagnostics when startupMemoryCache is enabled
- restore memory bytes only after emscripten's emscripten_environ_constructor

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
measure

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

# Conflicts:
#	src/mono/wasm/runtime/startup.ts

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

kg commented on src/mono/sample/wasm/browser-advanced/index.html:
```diff
@@ -11,8 +11,6 @@
   <script type='module' src="./dotnet.js"></script>
   <link rel="preload" href="./mono-config.json" as="fetch" crossorigin="anonymous">
   <link rel="prefetch" href="./dotnet.wasm" as="fetch" crossorigin="anonymous">
-  <link rel="prefetch" href="./icudt.dat" as="fetch" crossorigin="anonymous">
```
do we know how serious the negative impact of this is on first load? it's valuable on later loads, of course. it may be worth evaluating whether prefetches should be kept, since on later loads the assets are probably already in the browser cache so the prefetch won't do anything bad

pavelsavara replied:
It would keep it alive in the cache. Is that bad enough ?

pavelsavara replied:
will revert this

kg commented on src/mono/wasm/runtime/assets.ts:
```diff
@@ -38,6 +38,19 @@ const skipBufferByAssetTypes: {
+} = {
+    "resource": true,
+    "assembly": true,
+    "pdb": true,
```
are memory snapshots enabled when debugging? should they be?

pavelsavara replied:
Yes, they are.

kg commented on src/mono/wasm/runtime/assets.ts:
```diff
@@ -38,6 +38,19 @@ const skipBufferByAssetTypes: {
     "dotnetwasm": true,
 };
 
+const skipInSnapshotByAssetTypes: {
```
does this suppress loading or delay it? looking at the code below, these bools determine whether assets go into a 'before' or 'after' list, so 'skipInSnapshot' is probably the wrong name for it. `loadAfterSnapshotByAssetTypes` maybe?

pavelsavara replied:
good feedback, renamed to `containedInSnapshotByAssetTypes` and collections to `alwaysLoadedAssets` and `containedInSnapshotAssets`

kg commented on src/mono/wasm/runtime/jiterpreter-jit-call.ts:
```diff
@@ -269,8 +269,9 @@ export function mono_jiterp_do_jit_call_indirect (
-    let failed = !getIsWasmEhSupported();
-    if (!failed) {
+    let failed = false;
+    const enabled = !runtimeHelpers.storeMemorySnapshotPending && getIsWasmEhSupported();
```
This code path will result in mono_jiterp_do_jit_call_indirect running for every JIT call, which may not be what we want? 

Is it the case that once the snapshot is taken, storeMemorySnapshotPending will become false and then the next call to this will run the initialization logic? Or is it possible that it could stay true forever on the first run?

kg replied:
To be clear, the problem is that `mono_jiterp_do_jit_call_indirect` is designed to run exactly once ever, and it's not fast. But by setting both `enabled` and `failed` to false, we ensure that it will keep getting called.

If we were to `cwraps.mono_jiterp_update_jit_call_dispatcher(0);` it would force the use of the C++ implementation instead, which is slower than indirect_js but faster than this.

pavelsavara replied:
I only want to temporarily disable it until after snapshot is taken. It should be able to execute (un-optimized) before that.

pavelsavara replied:
Now I see that it's only one function pointer. I will try to solve it.

pavelsavara replied:
Solved, thanks for this.

kg commented on src/mono/wasm/runtime/run.ts:
```diff
@@ -118,8 +118,10 @@ function set_exit_code_and_quit_now(exit_code: number, reason?: any): void {
+        else if (!reason) {
             reason = new runtimeHelpers.ExitStatus(exit_code);
+        } else {
+            exit_code = reason.status;
```
if `reason` is `any`, it may have a non-numeric `reason.status`

kg commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -295,14 +319,15 @@ async function postRunAsync(userpostRun: (() => void)[]) {
     afterOnRuntimeInitialized.promise_control.reject(reason);
     afterPostRun.promise_control.reject(reason);
-    if (should_exit) {
+    if (should_exit && reason.silent !== true) {
```
if `reason` is `any` then the expression `reason.silent` could throw an error

kg commented on src/mono/wasm/runtime/storage.ts:
```diff
@@ -0,0 +1,170 @@
+        if (!inputsHash) {
+            return undefined;
+        }
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${inputsHash}`;
```
why do we duplicate the creation of `cacheKey`, `openCache` call, and `cache.match` in these two functions? Is it expected that we will call `getMemorySnapshotSize` and then `getMemorySnapshot` sequentially in a single run?
I have mild concern about the duplication allowing bugs to creep in, but also that we could end up paying some costs to load the snapshot from cache twice (once to get the size, then to actually get it) since they are separate response objects.

pavelsavara replied:
moved into `getCacheKey`

kg commented on src/mono/wasm/runtime/storage.ts:
```diff
@@ -0,0 +1,170 @@
+    delete inputs.pthreadPoolSize;
+    delete inputs.asyncFlushOnExit;
+    delete inputs.assemblyRootFolder;
+    delete inputs.remoteSources;
```
Not your problem to fix necessarily, but SHA integrity is probably compromised if you ever loaded anything from a remote source, since the remote source could deploy a new version of it at any time.
We might want to consider disabling snapshots if remote sources are in use at all. (Do customers use them? The intent was to allow hosting ICU data on a CDN, but I don't know if anyone ever did it.)

pavelsavara replied:
`remoteSources` needs to comply with the `integrity` of the asset. 
We are adding them to `fetch` request already. 
Otherwise the download itself would fail.
That probably means that people would need to have namespaces in their CDN url.

kg commented on src/mono/wasm/runtime/storage.ts:
```diff
@@ -0,0 +1,170 @@
+    delete inputs.diagnosticTracing;
+    delete inputs.appendElementOnExit;
+    delete inputs.logExitCode;
+    delete inputs.pthreadPoolSize;
```
If the pthreadPoolSize changes between runs, could that cause problems with stuff inside the snapshot that encoded the pool size?

pavelsavara replied:
Probably not as all threads are started later. I tested with threads locally.

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -243,53 +242,95 @@ const App = {
+        }
+
+        // this is subsequent run with the actual tests. It will use whatever was cached in the previous run. 
+        // This way, we are testing that the cached version works.
```
I'm fine with this in principle, but it does seem like down the road we could run into issues where all of our automated tests are relying on the behavior of a loaded snapshot, and we end up with first-run-only bugs. I'm not sure if there's anything you can do about that though.

pavelsavara replied:
There  are lot of tests running on V8 and NodeJS. They don't get this change.

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -243,53 +242,95 @@ const App = {
+        console.log("Application arguments: " + runArgs.applicationArguments.join(' '));
+
+        if (is_browser) {
+            const dryOk = await dry_run(runArgs);
```
Just to verify: This will *always* create a snapshot right? It won't reuse a matching one it happens to find in the browser cache if (for example) someone is re-running tests in a browser tab with the debugger open to troubleshoot?

To me `exitAfterSnapshot: true` implies 'always make a new snapshot and save it' but I'm not certain from reading everything whether that's the intent and behavior here.

pavelsavara replied:
I added comment.

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -243,53 +242,95 @@ const App = {
+        dotnet.withRuntimeOptions(runArgs.runtimeArgs);
+    }
+    if (runArgs.debugging) {
+        dotnet.withDebugging(-1);
```
How should `withDebugging` and `withWaitingForDebugger` behave for the `dry_run` step? Wouldn't the dry run always hang if it's waiting for debugger, or is it set up so that the create finishes and loading the snapshot will resume waiting for it?

pavelsavara replied:
I guess this is unit test only internal thing, but maybe part of the debugger tests are failing because of that.
I will disable snapshot if I find `waitForDebugger` set. 

The `withDebugging` is just state of mono initialization with no outside dependency, so it should be OK to snapshot.
I will do more testing.

pavelsavara committed:
fix

pavelsavara committed:
optimize timezone detection

pavelsavara committed:
feedback

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix

pavelsavara committed:
- enable snapshot in blazor

 - @kg feedback about mono_jiterp_do_jit_call_indirect

pavelsavara committed:
store cache key

pavelsavara committed:
feedback and cleanup

pavelsavara committed:
fix debugging

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3

pavelsavara commented:
CI failures 
- `System.Drawing.PrimitivesTest.RectangleTests.RectangleFConversionTest ` is https://github.com/dotnet/runtime/issues/60585 unrelated
- `NuGet-Migrations` is unrelated and known
- Multi-threaded `DebuggerTests` failures are unrelated
- `System.Net.WebSockets.Client.Tests` timeouts on nodeJS is unrelated

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+# Memory snapshot of the Mono runtime #
```
```suggestion
# Memory snapshot of the Mono runtime
```

maraf commented on src/mono/wasm/memory-snapshot.md:
```diff
@@ -0,0 +1,27 @@
+
+We take snapshot of WASM memory after first cold start at run time​.
+We store it on the client side in the browser cache.
+For subsequent runs with the same configuration and same assets, we use the snapshot .
```
```suggestion
For subsequent runs with the same configuration and same assets, we use the snapshot
```

maraf commented on src/mono/wasm/runtime/assets.ts:
```diff
@@ -73,14 +88,51 @@ export async function mono_download_assets(): Promise<void> {
             }
+        };
+
+        // start fetching and assets in parallel, only assets which are not part of memory snapshot
```
```suggestion
        // start fetching assets in parallel, only assets which are not part of memory snapshot
```

maraf commented on src/mono/wasm/runtime/assets.ts:
```diff
@@ -73,14 +88,51 @@ export async function mono_download_assets(): Promise<void> {
+        // continue after we know if memory snapshot is available or not
+        await memorySnapshotSkippedOrDone.promise;
+
+        // start fetching and assets in parallel, only if memory snapshot is not available.
```
```suggestion
        // start fetching assets in parallel, only if memory snapshot is not available.
```

maraf reviewed (approved):
Looks good to me 👍

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/memory-snapshot.md

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/runtime/assets.ts

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/mono/wasm/runtime/assets.ts

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Merge branch 'main' into browser_memory_snapshot3
